File: egghead.io/data-structures-and-algorithms-in-js/lesson-10-build-a-bt-in-js-and-traversal/binary_tree.py
from node import Node


class BinaryTree:
  def __init__(self, data):
    self.root = Node(data)

  def traverse(self, node, func, order):
    if node:
      if order == 'in':
        self.traverse(node.prev, func, 'in')
        func(node)
        self.traverse(node.succ, func, 'in')
      elif order == 'pre':
        func(node)
        self.traverse(node.prev, func, 'pre')
        self.traverse(node.succ, func, 'pre')
      elif order == 'post':
        self.traverse(node.prev, func, 'post')
        self.traverse(node.succ, func, 'post')
        func(node)

File: egghead.io/data-structures-and-algorithms-in-js/lesson-10-build-a-bt-in-js-and-traversal/binary-tree.js
const Node = require('./node');


class BinaryTree {
  constructor(data) {
    this.root = new Node(data);
  }

  traverse(node, func, order) {
    if (node !== null) {
      switch (order) {
        case 'in': {
          this.traverse(node.prev, func, 'in'); 
          func(node);
          this.traverse(node.next, func, 'in');
          break;
        }
        case 'pre': {
          func(node);
          this.traverse(node.prev, func, 'pre');
          this.traverse(node.next, func, 'pre');
          break;
        }
        case 'post': {
          this.traverse(node.prev, func, 'post');
          this.traverse(node.next, func, 'post');
          func(node);
          break;
        }

        // no default
      }
    }
  }
}

module.exports = BinaryTree;
